File: elixir/lib/rational.ex
defmodule Rational do
  defstruct [:n, :d]

  def sign(x) when x>0,do: 1
  def sign(x) when x<0,do: -1
  def sign(0),do: 0

  def new(_,0) , do: :nan
  def new(n,d) do
    g = (sign d) * (gcd abs(n), abs(d))

    %Rational{n: Kernel.div(n,g), d: Kernel.div(d,g)}
  end

  def from_int(n), do: new n,1

  def gcd(a,0), do: a
  def gcd(a,b), do: gcd b, rem(a,b)

  def to_string(:nan),do: "NaN"
  def to_string(%Rational{n: n, d: 1}), do: Kernel.to_string n
  def to_string(%Rational{n: n, d: d}), do: Kernel.to_string(n) <> "/" <> Kernel.to_string(d)

  def add(%Rational{n: a_n, d: a_d},%Rational{n: b_n, d: b_d}), do: new(a_n * b_d + b_n * a_d, a_d * b_d)
  def add(_,_), do: :nan

  def sub(%Rational{n: a_n, d: a_d},%Rational{n: b_n, d: b_d}), do: new(a_n * b_d - b_n * a_d, a_d * b_d)
  def sub(_,_), do: :nan

  def mul(%Rational{n: a_n, d: a_d},%Rational{n: b_n, d: b_d}), do: new(a_n*b_n,a_d*b_d)
  def mul(_,_), do: :nan

  def div(%Rational{n: a_n, d: a_d},%Rational{n: b_n, d: b_d}), do: new(a_n*b_d,a_d*b_n)
  def div(_,_), do: :nan
end

File: js/rational.ts
export type Rational = {
  n: number, d: number
} | null;

function gcd(a: number, b: number): number {
  if (b === 0) {
    return a;
  } else {
    return gcd(b, a % b)
  }
}

export function rational(n: number, d: number) {
  if (d == 0) {
    return null;
  } else {
    let g = gcd(Math.abs(n), Math.abs(d));
    g = d < 0 ? -g : g;
    return {
      n: n / g | 0,
      d: d / g | 0
    };
  }
}


export function toString(self: Rational) {
  if (self === null) {
    return "NaN";
  } else {
    if (self.d === 1) {
      return self.n.toString();
    } else {
      return `${self.n}/${self.d}`;
    }
  }
}

export function add(a: Rational, b: Rational) {
  if (a !== null && b !== null) {
    return rational(a.n * b.d + b.n * a.d, a.d * b.d);
  } else {
    return null;
  }
}

export function sub(a: Rational, b: Rational) {
  if (a !== null && b !== null) {
    return rational(a.n * b.d - b.n * a.d, a.d * b.d);
  } else {
    return null;
  }
}

export function mul(a: Rational, b: Rational) {
  if (a !== null && b !== null) {
    return rational(a.n * b.n, a.d * b.d);
  } else {
    return null;
  }
}

export function div(a: Rational, b: Rational) {
  if (a !== null && b !== null) {
    return rational(a.n * b.d, a.d * b.n);
  } else {
    return null;
  }
}